src/test/java/com/google/sps/TripServletTest.java

```diff
@@ -0,0 +1,274 @@
+      tripName, destinationName, tripDayOfTravel, photoSrc);
+
+    // Retrieve the datastore results.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query(Trip.TRIP);
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> listResults = results.asList(FetchOptions.Builder.withDefaults());
+
+    // Check whether the proper Entity and count were returned.
+    Assert.assertEquals(1, listResults.size());
+    Assert.assertEquals(tripName, listResults.get(0).getProperty(Trip.TRIP_NAME));
+    Assert.assertEquals(destinationName, listResults.get(0).getProperty(Trip.DESTINATION_NAME));
+    Assert.assertEquals(tripDayOfTravel, listResults.get(0).getProperty(Trip.START_DATE));
+    Assert.assertEquals(tripDayOfTravel, listResults.get(0).getProperty(Trip.END_DATE));
+    Assert.assertEquals(photoSrc, listResults.get(0).getProperty(Trip.IMAGE_SRC));
+    
+    // Confirm that the Entity in the database matches the method return.
+    Assert.assertEquals(listResults.get(0), tripEntityReturn);
+  }
+
+  @Test
+  public void testStoreTripEntityLoggedInCheckUser() throws Exception {
+    // Mock response.      
+    HttpServletResponse responseMock = mock(HttpServletResponse.class);
+    
+    // Create Trip Entity properties.
+    final String tripName = "Family Vacation";
+    final String destinationName = "Island of Hawai'i";
+    final String tripDayOfTravel = "2020-07-17";
+    final String photoSrc = "../images/placeholder_image.png";
+
+    // Mock UserService methods as logged-in user.
+    UserService userServiceMock = mock(UserService.class);
+    when(userServiceMock.isUserLoggedIn()).thenReturn(true);
+    // This is the User object from Google Appengine (full path given to avoid
+    // confusion with local User.java file).
+    when(userServiceMock.getCurrentUser()).thenReturn(
+        new com.google.appengine.api.users.User(EMAIL, AUTH_DOMAIN));
+    when(userServiceMock.createLogoutURL(AuthServlet.redirectUrl)).thenReturn(LOGOUT_URL);
+
+    // PowerMock static getUserService() method, which is used to get the user.
+    PowerMockito.mockStatic(UserServiceFactory.class);
+    when(UserServiceFactory.getUserService()).thenReturn(userServiceMock);
+
+    // Run storeTripEntity(...), with the User logged in (so trip is stored).
+    tripServlet.storeTripEntity(responseMock, tripName, destinationName, 
+      tripDayOfTravel, photoSrc);
+
+    // Retrieve the datastore results.
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    Query query = new Query(Trip.TRIP);
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> listResults = results.asList(FetchOptions.Builder.withDefaults());
+
+    // Check that the Trip Entity has the User as an ancestor / parent.
+    Key userAncestorKey = listResults.get(0).getParent();
+    Entity userEntity = datastore.get(userAncestorKey);
```

Comment:
nice! this is really useful in a test b/c, in addition to testing the functionality, reading this clarifies how Trip and User entities are related to one another

Reply:
Thanks! :)